packages/react/src/ReferenceInput/ReferenceInput.tsx

```diff
@@ -16,75 +17,253 @@ export interface ReferenceInputProps {
   onChange?: (value: Reference | undefined) => void;
 }
 
+interface BaseTargetType {
+  value: string;
+}
+
+type ProfileTargetType = BaseTargetType & {
+  type: 'profile';
+  name?: string;
+  title?: string;
+  resourceType?: string;
+  error?: any;
+};
+
+type ResourceTypeTargetType = BaseTargetType & {
+  type: 'resourceType';
+  resourceType: string;
+};
+type TargetType = ResourceTypeTargetType | ProfileTargetType;
+
 export function ReferenceInput(props: ReferenceInputProps): JSX.Element {
-  const targetTypes = getTargetTypes(props.targetTypes);
-  const initialResourceType = getInitialResourceType(props.defaultValue, targetTypes);
+  const medplum = useMedplum();
   const [value, setValue] = useState<Reference | undefined>(props.defaultValue);
-  const [resourceType, setResourceType] = useState<ResourceType | undefined>(initialResourceType);
+  const [targetTypes, setTargetTypes] = useState<TargetType[] | undefined>(() => createTargetTypes(props.targetTypes));
+  const [state, setState] = useState<{
+    targetType?: TargetType;
+    resourceType?: ResourceType;
```

Comment:
Is `resourceType` always `targetType.resourceType`?  If so, then might be more straightforward to make it a derived value.

Reply:
I've been going back and forth on this...

There is one case where `resourceType !== targetType.resourceType`: When `props.targetTypes` isn't specified,  `state.targetType === undefined` and the `<ResourceTypeInput />` mode of this component is used which is totally based on `resourceType`  and avoids `targetTypes` and `targetType` altogether similar to how this component worked before this PR.

I just pushed a commit that just adds a wrapper `TargetType` for use in the `<ResourceTypeInput />` mode to make the simplification you suggest and I don't hate it!